### PR TITLE
TSK-621 - Make classification categories configurable by type

### PR DESCRIPTION
**Pull request by ignaciorecuerda**
Join ClassificationTypesService and ClassificationCategoriesService.
Remove ClassificationTypesService.
Add default icon and deselect classification when changing the category
Create CustomFieldsService test

### CODE REVIEW

File: lib/taskana-core/src/test/resources/taskana.properties [hunk 1]
```diff
@@ -6,7 +6,7 @@ taskana.roles.monitor=john|teamlead_2 | monitor
 taskana.domains= Domain_A  ,  DOMAIN_B
 
 taskana.classification.types= TASK , document
-taskana.classification.categories= EXTERNAL , manual, autoMAtic  ,Process
+taskana.classification.categories= TASK: EXTERNAL , manual, autoMAtic, Process; document: Default;
```
Comment (holgerhagen) [inline lib/taskana-core/src/test/resources/taskana.properties]: I would prefer the following configuration:
taskana.classification.categories.TASK= EXTERNAL, ...
taskana.classification.categories.DOCUMENT= EXTERNAL, ...
What do you think?

File: rest/taskana-rest-spring/pom.xml [hunk 2]
```diff
@@ -214,7 +214,12 @@
                             <groupId>commons-logging</groupId>
                         </exclusion>
                     </exclusions>
-                </dependency>                
+                </dependency>
```
Comment (holgerhagen) [inline rest/taskana-rest-spring/pom.xml]: Why is this required? We already have this dependency:

                <dependency>
                        <groupId>pro.taskana</groupId>
                        <artifactId>taskana-spring</artifactId>
                        <version>${project.version}</version>
                </dependency>

File: lib/taskana-core/src/main/java/pro/taskana/impl/ClassificationServiceImpl.java [hunk 3]
```diff
@@ -290,7 +290,7 @@ private void initDefaultClassificationValues(ClassificationImpl classification)
         }
 
         if (classification.getCategory() != null
-            && !taskanaEngine.getConfiguration().getClassificationCategories().contains(classification.getCategory())) {
+                && !taskanaEngine.getConfiguration().getClassificationCategories(classification.getType()).contains(classification.getCategory())) {
             throw new InvalidArgumentException("Given classification category " + classification.getCategory()
```
[the remark this answers is not in the document]
Comment (ignaciorecuerda) [reply lib/taskana-core/src/main/java/pro/taskana/impl/ClassificationServiceImpl.java]: done

File: lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java [hunk 4]
```diff
@@ -94,7 +94,7 @@
     protected List<String> classificationTypes = new ArrayList<String>();
 
     // List of configured classification categories
```
[the remark this answers is not in the document]
Comment (ignaciorecuerda) [reply lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java]: why not? I think the unique change is in classificationCategories (the new hashMap) but I think the type of classificationTypes should still be a List<String>, is it?

[the remark this answers is not in the document]
Comment (ignaciorecuerda) [reply lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java]: I remove the comment and rename to classificationCategoriesByType

File: web/src/app/administration/services/classification-categories-service/classification-categories.service.ts [hunk 5]
```diff
@@ -2,34 +2,41 @@ import { HttpClient } from '@angular/common/http';
...
+import { Observable ,  ReplaySubject, BehaviorSubject } from 'rxjs';
 import { CustomFieldsService } from 'app/services/custom-fields/custom-fields.service';
 import { Pair } from 'app/models/pair';
 
 @Injectable()
 export class ClassificationCategoriesService {
 
-  private url = environment.taskanaRestUrl + '/v1/classification-categories';
-  private dataObs$ = new ReplaySubject<Array<string>>(1);
+  private domain = environment.taskanaRestUrl;
```
[the remark this answers is not in the document]
Comment (ignaciorecuerda) [reply web/src/app/administration/services/classification-categories-service/classification-categories.service.ts]: Done!

File: web/src/app/administration/services/classification-categories-service/classification-categories.service.ts [hunk 6]
```diff
@@ -2,34 +2,41 @@ import { HttpClient } from '@angular/common/http';
...
-    if (!this.dataObs$.observers.length || forceRefresh) {
-      this.httpClient.get<Array<string>>(this.url).subscribe(
-        data => { this.dataObs$.next(data); this.categoriesObject = this.getCustomCategoriesObject(data) },
-        error => {
-          this.dataObs$.error(error);
-          this.dataObs$ = new ReplaySubject(1);
-        }
+  getCategories(type?: string): Observable<Array<string>> {
+    if (!this.dataObsCategories$.observers.length || type !== this.type) {
+      this.httpClient.get<Array<string>>(!type ? this.urlCategories : this.urlCategories + this.param + type).subscribe(
```
[the remark this answers is not in the document]
Comment (ignaciorecuerda) [reply web/src/app/administration/services/classification-categories-service/classification-categories.service.ts]: Done!

File: rest/taskana-rest-spring/pom.xml (hunk 2 above)
Comment (ignaciorecuerda) [reply rest/taskana-rest-spring/pom.xml]: you are right

File: lib/taskana-core/src/test/resources/taskana.properties (hunk 1 above)
Comment (ignaciorecuerda) [reply lib/taskana-core/src/test/resources/taskana.properties]: Your way is scalable. Thank you for your advice

File: web/src/app/services/custom-fields/taskana-customization-test.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+                  "field": "Classification custom 1",
+                  "visible": true
+              },
+              "custom3": {
+                  "field": "",
+                  "visible": false
+              }
+          },
+          "categories": {
+              "EXTERNAL": "assets/icons/categories/external.svg",
```
Comment (holgerhagen) [inline web/src/app/services/custom-fields/taskana-customization-test.json]: Isn't there DEFAULT missing?

File: lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java [hunk 8]
```diff
@@ -417,12 +425,19 @@ public void setClassificationTypes(List<String> classificationTypes) {
         this.classificationTypes = classificationTypes;
     }
 
-    public List<String> getClassificationCategories() {
+    public List<String> getClassificationCategoriesByType() {
```
Comment (holgerhagen) [inline lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java]: This should rather be names getAllClassificationCategories, shouldn't it?

File: lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java [hunk 9]
```diff
@@ -93,8 +93,7 @@
     // List of configured classification types
     protected List<String> classificationTypes = new ArrayList<String>();
 
-    // List of configured classification categories
-    protected List<String> classificationCategories = new ArrayList<String>();
+    protected Map<String, List<String>> classificationCategoriesByType = new HashMap<String, List<String>>();
```
Comment (holgerhagen) [inline lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java]: please name it classificationCategoriesByTypeMap. Sorry. But this helps me to understand the code more easily. Thank you.

File: web/src/app/services/custom-fields/taskana-customization-test.json (hunk 7 above)
Comment (ignaciorecuerda) [reply web/src/app/services/custom-fields/taskana-customization-test.json]: No, it isn't necessary. In this file, we only put the custom icons. But if you prefer I can put the default icon

File: lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java (hunk 8 above)
Comment (ignaciorecuerda) [reply lib/taskana-core/src/main/java/pro/taskana/configuration/TaskanaEngineConfiguration.java]: yes, you are right. This method returns all the classification categories.
Thank you
